File: SymbTab.java
/**
	Membros do grupo:
	Herbert Alves Batista
	Luiz Gustavo Avelino Mendes
	Thiago Henrique Balbino Dias
 */

import java.util.Hashtable;

public class SymbTab{

	//a tabela de simbolos será estruturada em uma tabela hash
	public static Hashtable<String, String> tabelaSimb;	

	public static final byte CONST = 1;
	public static final byte INTEGER = 2;
	public static final byte BYTE = 3;
	public static final byte STRING = 4;
	public static final byte WHILE = 5;
	public static final byte IF = 6;
	public static final byte ELSE = 7;
	public static final byte AND = 8;
	public static final byte OR = 9;
	public static final byte NOT = 10;
	public static final byte EQUAL = 11; // sinal -> ==
	public static final byte ASSIG = 12; // sinal -> =
	public static final byte RPARENT = 13; // abre parenteses
	public static final byte LPARENT = 14; // fecha parenteses
	public static final byte MAIOR = 15; // sinal -> >
	public static final byte MENOR = 16; // sinal -> <
	public static final byte DIFERENTE = 17; // sinal -> <>
	public static final byte MAIORIGUAL = 18; // sinal -> >=
	public static final byte MENORIGUAL = 19; // sinal -> <=
	public static final byte COMMA = 20; //virgula
	public static final byte MAIS = 21; // sinal -> +
	public static final byte MENOS = 22; // sinal -> -
	public static final byte MULT = 23; // sinal -> *
	public static final byte DIV = 24; // sinal -> /
	public static final byte SEMICOLON = 25; // sinal -> ;
	public static final byte RCHAVE = 26; // sinal -> {
	public static final byte LCHAVE = 27; // sinal -> }
	public static final byte READLN = 28;
	public static final byte WRITE = 29;
	public static final byte WRITELN = 30;
	public static final byte TRUE = 31;
	public static final byte FALSE = 32;
	public static final byte BOOLEAN = 33;


	/**
		metodo que inicializa a tabela de simbolos com 
		as palavras reservadas e identificadores
	*/
	public static void inicialize (){
		tabelaSimb = new Hashtable<String, String>(); 

//                  (chave, lexema)
		insertSymb (CONST, "const");
		insertSymb (INTEGER, "integer");
		insertSymb (BYTE, "byte");
		insertSymb (STRING, "string");
		insertSymb (WHILE, "while");
		insertSymb (IF, "if");
		insertSymb (ELSE, "else");
		insertSymb (AND, "and");
		insertSymb (OR, "or");
		insertSymb (NOT, "not");
		insertSymb (EQUAL, "==");
		insertSymb (ASSIG, "=");
		insertSymb (RPARENT, "(");
		insertSymb (LPARENT, ")");
		insertSymb (MENOR, "<");
		insertSymb (MAIOR, ">");
		insertSymb (DIFERENTE, "<>");
		insertSymb (MAIORIGUAL, ">=");
		insertSymb (MENORIGUAL, "<=");
		insertSymb (COMMA, ",");
		insertSymb (MAIS, "+");
		insertSymb (MENOS, "-");
		insertSymb (MULT, "*");
		insertSymb (DIV, "/");
		insertSymb (SEMICOLON, ";");
		insertSymb (RCHAVE, "{");
		insertSymb (LCHAVE, "}");
		insertSymb (READLN, "readln");
		insertSymb (WRITE, "write");
		insertSymb (WRITELN, "writeln");
		insertSymb (TRUE, "true");
		insertSymb (FALSE, "false");
		insertSymb (BOOLEAN, "boolean");

	}//end inicialize


	/**
		metodo que inserte chave(key) e lexema(lex) na tabela de simbolos
		@param key: chave de pesquisa na tabela de simbolos
		@param lex: lexema que sera inserido na tabela de simbolos	
	*/
	public static void insertSymb (byte key, String lex){		

		String chave = toString(key); //convertendo key para string
		String min = lex.toLowerCase(); //passando lexema para minusculo
		tabelaSimb.put(chave, min); //gravando key e lex na tabela de simbolos

	}//end insertSymb


	/**
	 	metodo que busca o lexema de um token
	 	@param key: chave que se refere ao token
	 	@return: lexema encontrado
	 */
	public static String getLexToken (byte key){				
		return (tabelaSimb.get(toString(key)));
	}//end getKeyToken



	/**
      Funcao que transforma byte em string
      @param num: byte a ser alterado
      @return : valor alterado para string
    */
    private static String toString ( byte num ){
        return (num + "");
    }//end isNumber



    /**
     	funcao de testes
     */
    public static void main(String[] args) {    
    	System.out.println("teste funcionando.");
    	inicialize();
    	
    	//for (byte i = 1; i <= 33; i++){
    	//	System.out.println ("teste = "+getLexToken(i));
    	//}

    	//String teste = toString(x);
    	//System.out.println (teste);	    	
    	
    }

}//end class

//.ao finalizar a classe SymbTab.java retirar os campos static dos metodos

File: SymbTab.cpp
/*
	Membros do grupo:
	Herbert Alves Batista
	Luiz Gustavo Avelino Mendes
	Thiago Henrique Balbino Dias
*/

//----------------------------------
// BIBLIOTECAS
//----------------------------------

#include <stdio.h>
#include <iostream>
#include <string>
#include "SymbTab.h"
using namespace std;

//----------------------------------
// IMPLEMENTACAO CLASSE SYMBTAB
//----------------------------------

/**
*   Construtor
*/
SymbTab::SymbTab(){
    tabSimb = new SymbolNode[SIZE];
    for(int i = 0; i < SIZE; i++){
        tabSimb[i].cod = 0;
        tabSimb[i].lexema = "";
		tabSimb[i].next = NULL;
    }
    inicialize();
}// end construtor()


/**
*   Destrutor
*/
SymbTab::~SymbTab(){
}//end destrutor()


/**
*	Metodo que inicializa a tabela de simbolos com
*	os codigos de cada token, e os lexemas das palavras reservadas.
*/
void SymbTab::inicialize(){
	insert('5', "while");
	insert('3', "if");
	insert('4', "else");
	insert('J', "and");
	insert('J', "or");
	insert('J', "not");
	insert('6', "readln");
	insert('7', "write");
	insert('7', "writeln");
	insert('L', "TRUE");
	insert('L', "FALSE");
	insert('1', "=");
	insert('2', "integer");
	insert('2', "byte");
	insert('2', "string");
	insert('2', "boolean");
	insert('8', "(");
	insert('9', ")");
	insert('A', ">");
	insert('A', "<");
	insert('A', "<>");
	insert('A', ">=");
	insert('A', "<=");
	insert('A', "==");
	insert('B', ",");
	insert('K', ";");
	insert('C', "+");
	insert('C', "-");
	insert('C', "*");
	insert('C', "/");
	insert('D', "{");
	insert('E', "}");
	insert('G', "const");
	insert ('P', "then");
}// end inicialize()


/**
 *  Metodo que recebe um token encontrado pelo analisador Lexico
 *  e o insere na tabela de simbolos e registro lexico
 *//*
 void SymbTab::input(byte cod, std::string lexema){
     SymbolNode * symbol = insert(cod, lexema);     // Insere o token na tabela de simbolos
     addToken(symbol);                              // Insere o token no registro lexico
 }*/


/**
*   Metodo que insere um token na tabela de simbolos
*/
SymbolNode * SymbTab::insert(byte cod, std::string lexema){

	int line = spreading(lexema); 	                // Armazena a posicao na tabela de simbolos
	SymbolNode *symbol;                     		// Simbolo a ser inserido na tabela de simbolos


	symbol = search(lexema);                        // Verifica se o lexema ja existe na tabela de simbolo

	if(symbol == NULL){ // token nao existe na tabela de simbolo

        symbol = new SymbolNode;
        symbol->cod = cod;
        symbol->lexema = lexema;
        symbol->next = NULL;

        SymbolNode * node = tabSimb[line].next;	// Auxiliar para caminhar na lista
        if(node == NULL){	// tabela vazia
            tabSimb[line].next = symbol;        // insere o token na primeira posição válida
        } else {
            while(node->next != NULL){
				node = node->next;              //caminha ate encontrar o ultimo
            }
            node->next = symbol;                // insere o token no fim da lista
            symbol = node->next;                // armazena endereço de inserção para retorno
        }
        node = NULL;
    }
	return symbol;
} // end insert()


/**
*   Metodo que define a posicao de um lexema na tabela de simbolos
*   Funcao de espalhamento
*/
int SymbTab::spreading(std::string lexema){
    int term = 0;

    for(int i = 0; i < (int)lexema.length(); i++){
        term += lexema.at(i);
    }
    return term % SIZE;
} // end spreading()

/**
 *  Metodo que verifica a existência de um lexema na tabela de simbolos e o retorna
 */
 SymbolNode * SymbTab::search(std::string lexema){

     SymbolNode * resposta = NULL;
     SymbolNode * node = tabSimb[spreading(lexema)].next;

     while(node != NULL){
            if(lexema.compare(node->lexema) == 0){
                return node;
            }
            node = node->next;
		}
     return resposta;
 }

/**
*   Funcao para testar
*/
void SymbTab::exibir (){

    cout << "\n\nquant\tidToken\tlexema\n";
    int quantidade = 0;
    for (int i =0; i < SIZE; i++){

        struct SymbolNode * node = tabSimb[i].next;
        while(node != NULL){
            cout << ++quantidade << "\t" << node->cod << "\t" << node->lexema << "\n";
            node = node->next;
        } // end while
    }// end for
}// end exibir()



